From 7d511257a0863dc86b7a8872d7f86c446248bde2 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Tue, 19 Aug 2014 15:25:25 -0400
Subject: [tools] perf/machine: Add method to loop over threads and invoke handler

Message-id: <1408462094-14194-157-git-send-email-jolsa@redhat.com>
Patchwork-id: 88138
O-Subject: [PATCH RHEL7.1 BZ1131394 156/325] perf machine: Add method to loop over threads and invoke handler
Bugzilla: 1131394
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1131394
https://bugzilla.redhat.com/show_bug.cgi?id=1131394

upstream
========
commit 35feee19f9fda7447f51073b5be3f6d082b508f5
Author: David Ahern <dsahern@gmail.com>
Date: Sat Sep 28 13:12:58 2013 -0600

description
===========
Loop over all threads within a machine - including threads moved to the
dead threads list -- and invoked a function.

This allows commands to run some specific function on each thread (eg.,
dump statistics) yet hides how the threads are maintained within the
machine.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/util/machine.c b/tools/perf/util/machine.c
index fc14f9b..901397a 100644
--- a/tools/perf/util/machine.c
+++ b/tools/perf/util/machine.c
@@ -1393,3 +1393,26 @@ int machine__resolve_callchain(struct machine *machine,
        sample);
 
 }
+
+int machine__for_each_thread(struct machine *machine,
+        int (*fn)(struct thread *thread, void *p),
+        void *priv)
+{
+ struct rb_node *nd;
+ struct thread *thread;
+ int rc = 0;
+
+ for (nd = rb_first(&machine->threads); nd; nd = rb_next(nd)) {
+  thread = rb_entry(nd, struct thread, rb_node);
+  rc = fn(thread, priv);
+  if (rc != 0)
+   return rc;
+ }
+
+ list_for_each_entry(thread, &machine->dead_threads, node) {
+  rc = fn(thread, priv);
+  if (rc != 0)
+   return rc;
+ }
+ return rc;
+}
diff --git a/tools/perf/util/machine.h b/tools/perf/util/machine.h
index 5150d5e..d44c09b 100644
--- a/tools/perf/util/machine.h
+++ b/tools/perf/util/machine.h
@@ -166,4 +166,8 @@ void machines__destroy_kernel_maps(struct machines *machines);
 
 size_t machine__fprintf_vmlinux_path(struct machine *machine, FILE *fp);
 
+int machine__for_each_thread(struct machine *machine,
+        int (*fn)(struct thread *thread, void *p),
+        void *priv);
+
 #endif /* __PERF_MACHINE_H */
-- 
1.7.1